pyabc/epsilon/temperature.py

```diff
@@ -0,0 +1,502 @@
+            fallback = self.temperatures[t - 1] \
+                if t-1 in self.temperatures else np.inf
+            proposed_value = self.aggregate_fun(temps)
+            # also a value lower than 1.0 does not make sense
+            temperature = max(min(proposed_value, fallback), 1.0)
+
+        # record found value
+        self.temperatures[t] = temperature
+
+    def __call__(self,
+                 t: int) -> float:
+        return self.temperatures[t]
+
+
+class TemperatureScheme:
+    """
+    A TemperatureScheme suggests the next temperature value. It is used as
+    one of potentially multiple schemes employed in the Temperature class.
+    This class is abstract.
+    """
+
+    def __init__(self):
+        pass
+
+    def __call__(self,
+                 t: int,
+                 get_weighted_distances: Callable,
+                 max_nr_populations: int,
+                 pdf_norm: float,
+                 kernel_scale: str,
+                 prev_temperature: float,
+                 acceptance_rate: float):
+        pass
+
+
+class AcceptanceRateScheme(TemperatureScheme):
+    """
+    Try to keep the acceptance rate constant at a value of
+    `target_rate`. Note that this scheme will fail to
+    reduce the temperature sufficiently in later iterations, if the
+    problem's inherent acceptance rate is lower, but it has been
+    observed to give big feasible temperature leaps in early iterations.
+
+    Parameters
+    ----------
+    target_rate: float
+        The target acceptance rate to match.
+    """
+
+    def __init__(self, target_rate: float = 0.5):
+        self.target_rate = target_rate
+
+    def __call__(self,
+                 t: int,
+                 get_weighted_distances: Callable,
+                 max_nr_populations: int,
+                 pdf_norm: float,
+                 kernel_scale: str,
+                 prev_temperature: float,
+                 acceptance_rate: float):
+        # execute function (expensive if in calibration)
+        df = get_weighted_distances()
+
+        weights = np.array(df['w'], dtype=float)
+        pdfs = np.array(df['distance'], dtype=float)
+
+        # compute rescaled posterior densities
+        if kernel_scale == SCALE_LIN:
+            values = pdfs / pdf_norm
+        else:  # kernel_scale == SCALE_LOG
+            values = np.exp(pdfs - pdf_norm)
+
+        # to acceptance probabilities
+        values = np.minimum(values, 1.0)
+
+        # to pmf
+        # TODO: We currently use the weights from the previous iteration.
+        # It might be better to use weights based on `transition / prior`
+        # for the next proposal transition kernel.
+        weights /= np.sum(weights)
+
+        # objective function which we wish to find a root for
+        def obj(beta):
+            val = np.sum(weights * values**beta) - self.target_rate
+            return val
+
+        if obj(1) > 0:
+            # function is mon. dec., smallest possible value already > 0
```

Comment:
is " mon. dec." an usual abbreviation in math?

Reply:
monotonically decreasing, spelled it out.